diff --git a/node_modules/nel/lib/nel.js b/node_modules/nel/lib/nel.js
index f2391d6..05874d7 100755
--- a/node_modules/nel/lib/nel.js
+++ b/node_modules/nel/lib/nel.js
@@ -199,7 +199,9 @@ Session._command = paths.node;
  * @member {String[]}
  * @private
  */
-Session._args = ["--eval", server]; // --eval workaround
+// some useful flags for debugging, turn on if need be
+// Session._args = ["--polyglot", "--jvm", "--inspect", "--inspect.Suspend=false", "--experimental-options", "--inspect.Internal=true", "--eval", server]; // --eval workaround
+Session._args = ["--polyglot", "--jvm", "--eval", server]; // --eval workaround
 
 /**
  * Combination of a piece of code to be run within a session and all the
diff --git a/node_modules/nel/lib/server/main.js b/node_modules/nel/lib/server/main.js
index 6935d32..ca6b62a 100755
--- a/node_modules/nel/lib/server/main.js
+++ b/node_modules/nel/lib/server/main.js
@@ -44,6 +44,7 @@ var DEBUG = !!process.env.DEBUG;
 var log;
 var requester;
 var initialContext;
+var esprima = require('esprima');
 
 // Init IPC server
 init();
@@ -313,6 +314,98 @@ function inspect(object) {
     }
 }
 
-function run(code) {
+function parseDeclarations(declarations) {
+    let declarationCandidates = declarations.map(declaration => {
+        if(!declaration.hasOwnProperty('id')) {
+            return null;
+        }
+        const name = declaration['id']['name'];
+        if(!declaration['init'] ||
+            (declaration['init']['type'] !== 'Literal' && declaration['init']['type'] !== 'ArrowFunctionExpression')) {
+            return null;
+        }
+        return name;
+    });
+    return declarationCandidates.filter(e => e !== null);
+}
+
+function findNestedDeclarationKeys(parsedCode) {
+    let varDeclarations = [];
+    if(!parsedCode) {
+        return varDeclarations;
+    }
+
+    if(parsedCode['declarations'] !== undefined) {
+        varDeclarations.push(...parseDeclarations(parsedCode['declarations']));
+    }
+
+    if(Array.isArray(parsedCode)) {
+        parsedCode.forEach(object => {
+            varDeclarations.push(...findNestedDeclarationKeys(object));
+        })
+    } else if(typeof(parsedCode) === 'object') {
+        Object.keys(parsedCode).map(key => {
+            varDeclarations.push(...findNestedDeclarationKeys(parsedCode[key]));
+        })
+    }
+    return varDeclarations;
+}
+
+function run(rawCode) {
+    const POLYGLOT_MAGIC_PATTERN = '%polyglot ';
+
+    if(rawCode.startsWith(POLYGLOT_MAGIC_PATTERN)) {
+        const rawCodeSplit = rawCode.split('\n');
+        const firstLine = rawCodeSplit.shift();
+
+        const firstLineSplit = firstLine.split(' ');
+        const languageFromMagic = firstLineSplit[1];
+
+        const code = rawCodeSplit.join('\n');
+        const evalResultToBeReturned = Polyglot.eval(languageFromMagic, code);
+        console.debug('evaluation of actual code succeeded');
+
+        let varsToBeExported = []
+        if(languageFromMagic === 'js') {
+            parsedCode = esprima.parse(code);
+            console.debug('parsed code:');
+            console.debug(JSON.stringify(parsedCode, null, 2));
+            varsToBeExported = findNestedDeclarationKeys(parsedCode);
+        } else {
+            varsToBeExported = firstLineSplit.slice(2);
+        }
+        varsToBeExported = Array.from(new Set(varsToBeExported));
+        console.log('these will be exported', varsToBeExported);
+
+        let varExportCodeStatements = varsToBeExported.reduce(
+            (code, varIdentifier) => {
+                let newExportLine = '';
+                if (languageFromMagic === 'js') {
+                    newExportLine = `Polyglot.export('${varIdentifier}', ${varIdentifier})`;
+                } else if (languageFromMagic === 'ruby') {
+                    // attempt to export the var prefixed with the global var identifier '$'
+                    // as a workaround for https://github.com/oracle/truffleruby/issues/1695
+                    newExportLine = `Polyglot.export('${varIdentifier}', $${varIdentifier})`;
+                } else if (languageFromMagic === 'python') {
+                    newExportLine = `import polyglot;polyglot.export_value(${varIdentifier}, '${varIdentifier}')`;
+                } else if (languageFromMagic === 'R') {
+                    newExportLine = `export('${varIdentifier}', ${varIdentifier})`;
+                }
+
+                code.push(newExportLine);
+                return code;
+            }, []
+        );
+
+        const separator = languageFromMagic === 'js' ? ';' : '';
+        varExportCode = varExportCodeStatements.join(`${separator}\n`) + separator;
+        console.debug('varExportCode to be evaluated in the next step');
+        console.debug('--------------\n' + varExportCode + '\n--------------');
+        Polyglot.eval(languageFromMagic, varExportCode);
+        console.debug('evaluation of var export code succeeded');
+
+        return evalResultToBeReturned;
+    }
+
     return vm.runInThisContext(code);
 }
