diff --git a/node_modules/nel/lib/nel.js b/node_modules/nel/lib/nel.js
index f2391d6..15fe9a8 100755
--- a/node_modules/nel/lib/nel.js
+++ b/node_modules/nel/lib/nel.js
@@ -199,7 +199,9 @@ Session._command = paths.node;
  * @member {String[]}
  * @private
  */
-Session._args = ["--eval", server]; // --eval workaround
+// some useful flags for debugging, turn on if need be
+//Session._args = ["--polyglot", "--jvm", "--inspect", "--inspect.Suspend=false", "--experimental-options", "--inspect.Internal=true", "--eval", server]; // --eval workaround
+Session._args = ["--polyglot", "--jvm", "--eval", server]; // --eval workaround
 
 /**
  * Combination of a piece of code to be run within a session and all the
diff --git a/node_modules/nel/lib/server/main.js b/node_modules/nel/lib/server/main.js
index 6935d32..db0655b 100755
--- a/node_modules/nel/lib/server/main.js
+++ b/node_modules/nel/lib/server/main.js
@@ -44,6 +44,7 @@ var DEBUG = !!process.env.DEBUG;
 var log;
 var requester;
 var initialContext;
+var esprima = require('esprima');
 
 // Init IPC server
 init();
@@ -313,6 +314,110 @@ function inspect(object) {
     }
 }
 
-function run(code) {
-    return vm.runInThisContext(code);
+
+function defaultLanguageStatements(variableName = '__i') {
+    // globalVariables: returning actual array raises error (see: https://github.com/graalvm/graaljs/issues/173)
+    // TODO: replace string repr with actual arrays
+    return {
+        'python': {
+            'importVariable': `import polyglot; ${variableName} = polyglot.import_value("${variableName}")`,
+            'exportVariable': `import polyglot; polyglot.export_value(${variableName}, "${variableName}")`,
+            'globalVariables': `repr(list(globals().keys())).replace("'", '"')`,
+        },
+        'R': {
+            'importVariable': `${variableName} <- import("${variableName}")`,
+            'exportVariable': `export("${variableName}", ${variableName})`,
+            'globalVariables': `paste('["', paste(ls(all.names=TRUE), collapse='", "'), '"]', sep="")`,
+        },
+        'ruby': {
+            'importVariable': `$${variableName} = Polyglot.import("${variableName}")`,
+            'exportVariable': `Polyglot.export("${variableName}", $${variableName})`,
+            'globalVariables': 'require "json"; global_variables.map{ |varId| varId[1..-1] }.to_json',
+        },
+        'js': {
+            'importVariable': `${variableName} = Polyglot.import("${variableName}")`,
+            'exportVariable': `Polyglot.export("${variableName}", ${variableName})`,
+            'globalVariables': 'JSON.stringify(Object.keys(global))',
+        },
+    }
+}
+
+
+function loadGlobalVariables(language) {
+    // TODO: replace string repr with actual arrays (see: https://github.com/graalvm/graaljs/issues/173 and above)
+    return JSON.parse(Polyglot.eval(language, defaultLanguageStatements()[language]['globalVariables']));
+}
+
+
+function loadPolyglotBlacklist(language) {
+    let polyglotVariableBlacklist = null;
+    try {
+        polyglotVariableBlacklist = Polyglot.import('__polyglotVariableBlacklist');
+    } catch (e) {};
+
+    if (!polyglotVariableBlacklist) {
+        polyglotVariableBlacklist = {};
+    }
+
+    if (!polyglotVariableBlacklist[language]) {
+        let newPolyglotVariableBlacklist = loadGlobalVariables(language);
+        polyglotVariableBlacklist[language] = newPolyglotVariableBlacklist;
+        Polyglot.export('__polyglotVariableBlacklist', polyglotVariableBlacklist);
+    }
+
+    return polyglotVariableBlacklist[language];
+}
+
+
+function importPolyglotVariables(language) {
+    let variableList = null;
+    try {
+        variableList = Polyglot.import('__bindings');
+    } catch (e) {};
+
+    console.log('varlist: ', variableList);
+
+    if (!variableList) {
+        variableList = [];
+        Polyglot.export('__bindings', variableList)
+    }
+
+    console.log('varlist: ', variableList);
+
+    for (let i of variableList) {
+        Polyglot.eval(language, defaultLanguageStatements(i)[language]['importVariable']);
+    }
+}
+
+
+function exportPolyglotVariables(language) {
+    let polyglotVariableBlacklist = loadPolyglotBlacklist(language);
+
+    let newPolyglotVariables = loadGlobalVariables(language);
+    let newPolyglotVariablesWoBlacklisted = newPolyglotVariables.filter(x => !polyglotVariableBlacklist.includes(x));
+    for (let i of newPolyglotVariablesWoBlacklisted) {
+        Polyglot.eval(language, defaultLanguageStatements(i)[language]['exportVariable']);
+    }
+
+    Polyglot.export('__bindings', newPolyglotVariablesWoBlacklisted);
+}
+
+
+function run(rawCode) {
+    const POLYGLOT_MAGIC_PATTERN = '%polyglot ';
+
+    if(rawCode.startsWith(POLYGLOT_MAGIC_PATTERN)) {
+        const rawCodeSplit = rawCode.split('\n');
+        const languageFromMagic = rawCodeSplit[0].replace(POLYGLOT_MAGIC_PATTERN, '');
+        const code = rawCodeSplit.slice(1).join('\n');
+    }
+
 }
